### PR TITLE
[JEP-216] Extract localization support infrastructure into localization-support plugin

### PR DESCRIPTION
**Pull request by daniel-beck**
**Work in progress**: `localization-support` is not yet hosted or released. Some commits need to be reverted before this is merged.

Downstream from https://github.com/kohsuke/localizer/pull/16, https://github.com/stapler/stapler/pull/156, https://github.com/jenkinsci/jenkins/pull/3729, and https://github.com/daniel-beck/localization-support-plugin

Also update the resource files with the current state from Jenkins core, as they're being removed in the PR linked above.

Later revisions of this PR also support webapp resources, both from plugins and from core. Examples:

# Core webapp

## Before

before-core

## After

after-core

# Plugin webapp

## Before

before-plugin

## After

(with placeholder resource)

after-plugin

### CODE REVIEW

**Comment by LinuxSuRen** [comment]
`LocalizationFacet` and `I18nRequestDispatcher` exists for help resource files. But it's not the best solution.

**Comment by daniel-beck** [comment]
@LinuxSuRen Got it. I just was confused because the help files were still in core, as well as in this plugin, so I thought this worked when it did not. There's quite a lot of duplication going on.

**Comment by LinuxSuRen** [comment]
@daniel-beck Actually it's working partially. So I did not remove those files from the core.

**Comment by LinuxSuRen** [comment]
@daniel-beck We should figure out a way to solve a similar situation. I found the exception earlier. He should send the message if someone made a change of localization key or parameters.

File: core/src/main/resources/hudson/model/UpdateCenter/CoreUpdateMonitor/message_zh_CN.properties [hunk 1]
```diff
@@ -22,7 +22,7 @@
 
 Or\ Upgrade\ Automatically=\u6216\u81EA\u52A8\u5347\u7EA7
 NewVersionAvailable=\u65B0\u7248\u672C\u7684 Jenkins ({0}) \u53EF\u4EE5<a href="{1}">\u4E0B\u8F7D</a> \
-  (<a href="${changelog.url}">\u53D8\u66F4\u8BB0\u5F55</a>)\u3002
+  (<a href="{2}">\u53D8\u66F4\u8BB0\u5F55</a>)\u3002
```
Comment (LinuxSuRen) [inline core/src/main/resources/hudson/model/UpdateCenter/CoreUpdateMonitor/message_zh_CN.properties]: I need to fix this right now. There's somebody already met this issues.

File: src/main/java/io/jenkins/plugins/localization_zh_cn/LocalizationContributorImpl.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import hudson.PluginWrapper;
+import io.jenkins.plugins.localization.support.LocalizationContributor;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.net.URL;
+
+@Extension
+public class LocalizationContributorImpl extends LocalizationContributor {
+    @CheckForNull
```
Comment (oleg-nenashev) [inline src/main/java/io/jenkins/plugins/localization_zh_cn/LocalizationContributorImpl.java]: Move annotations to the localizer library?

**Comment by daniel-beck** [comment]
@LinuxSuRen Thanks for identifying this problem, you're right.

It seems I originally did that in https://github.com/jenkinsci/localization-support-plugin/commit/492bb166f5212ae8600fd83f2279f3097deab0c5 but removed it again in https://github.com/jenkinsci/localization-support-plugin/commit/a5b4e4f07278b816c92ed482fcb7709524b69ba0#diff-ba7f3e4f89111bd652dffe39ab80d9c7 because I didn't remember why it was needed.

Will fix in localization-support and integrate it here.

**Comment by daniel-beck** [comment]
@LinuxSuRen The PR linked in the latest commit should restore this functionality. Could you test the current state of this PR in depth? I expect all kinds localizations to be supported as it is, with the build of localization-support integrated in https://github.com/jenkinsci/localization-zh-cn-plugin/pull/31/commits/8cf2438744b2476e8d237c123b114c956086d974

**Comment by LinuxSuRen** [comment]
@daniel-beck Of course. I'll test this by manul.

**Pull request merged**